Optimize Massive Disconnect Scenario

larshesel opened the pull request:
I think this is ready for review /cc @dergraf

larshesel committed:
Share work over multiple supervisors

When many (say 100K) clients disconnect they'd overload the mailbox of
the single vmq_queue_sup supervisor. With this we share the load between
a configurable number of vmq_queue_sup supervisors.

larshesel committed:
Support otp17

Childspecs as maps was introduced in otp18.

larshesel committed:
Fix concurrency bug

larshesel committed:
Optimize start_queue/1 and get_queue_pid/1

We don't have to iterate over all the tables of the children supervisors
as we can map directly to the responsible supervisor using the phash2/1
function.

larshesel committed:
Avoid using slow supervisor:which_children/1

Register the child supervisor names instead and dynamically calculate
names when querying.

larshesel committed:
Small cleanup

larshesel committed:
Start persisted queues after sups are ready

larshesel committed:
Set default number of queue sups to 50

This should support most use cases up to about 200-300K queues, as each
sup would handle 4k-6k queues and would be able to handle the queue
teardown within a `reasonable time` (which of course depends on the
number of cores and how powerful they are).

larshesel commented:
@dergraf this is ready for review

dergraf commented on src/vmq_queue_sup.erl:
```diff
@@ -203,16 +161,16 @@ loop(State = #state{parent=Parent}, NrOfChildren) ->
     _ = [begin
              unlink(P),
              exit(P, kill)
          end || P <- get_keys(true)],
```
Instead of also storing the pids in the Process Dictionary, we could rely only on ETS. In this case we'd need to ets:fold, unlink, exit, and then ets:delete_all_objects.

Should we implement this as part of this PR?

dergraf commented on src/vmq_queue_sup.erl:
```diff
@@ -242,10 +200,10 @@ wait_children(NrOfChildren) ->
         {ok, Pid} ->
-            ets:insert(?QUEUE_TAB, {SubscriberId, Pid}),
+            ets:insert(QueueTab, {SubscriberId, Pid}),
             put(Pid, true),
```
Should we keep the Process Dictionary? See comment above

dergraf commented on src/vmq_queue_sup_sup.erl:
```diff
@@ -0,0 +1,156 @@
+    vmq_queue_sup:start_queue(SupName, SubscriberId, Clean).
+
+num_child_sups() ->
+    application:get_env(vmq_server, queue_sup_sup_children, 50).
```
We should expose this config value as a hidden config in cuttlefish

dergraf reviewed (commented):
In general, good to merge. We may want to address the mentioned issues in this PR too. What do you think.

larshesel commented:
I think I'll address all your comments in this PR. Should be quite easy to do. Will try to get it done tomorrow afternoon.

larshesel committed:
Don't store pids in process dict

They're redundant as we store them in the ets table as well.

Also fixed some badly indented code.

larshesel committed:
Add queue_sup_sup_children as hidden ctlfish param

larshesel commented:
@dergraf I think that it's all there now.

dergraf commented:
LGTM!
